src/lib/statistics/chunk_statistics/histograms/abstract_histogram.hpp

```diff
@@ -0,0 +1,214 @@
+ * Additionally, as of now, we only support a range of ASCII characters, that is stored as a string
+ * in `_supported_characters`. This range must not include gaps and the string has to be sorted.
+ * The possible maximum length of the prefix depends on the number of supported characters.
+ * It must not overflow the uint64_t data type used to represent strings as numbers.
+ * The formula used to verify the prefix length is:
+ * string_prefix_length < std::log(std::numeric_limits<uint64_t>::max()) / std::log(supported_characters.length() + 1)
+ */
+template <typename T>
+class AbstractHistogram : public AbstractFilter {
+ public:
+  AbstractHistogram();
+  AbstractHistogram(const std::string& supported_characters, const uint32_t string_prefix_length);
+  virtual ~AbstractHistogram() = default;
+
+  /**
+   * Strings are internally transformed to a number, such that a bin can have a numerical width.
+   * This transformation is based on uint64_t.
+   */
+  using HistogramWidthType = std::conditional_t<std::is_same_v<T, std::string>, uint64_t, T>;
+
+  virtual HistogramType histogram_type() const = 0;
+
+  /**
+   * Returns a string with detailed information about the histogram, including the edges of the individual bins.
+   */
+  std::string description() const;
+
+  /**
+   * Returns the estimated selectivity given a predicate type and its parameter(s).
+   * It will always be between 0 and 1.
+   */
+  float estimate_selectivity(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                             const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns the estimated cardinality given a predicate type and its parameter(s).
+   * It will always be between 0 and total_count().
+   * This method is specialized for strings to handle predicates uniquely applicable to string columns.
+   */
+  float estimate_cardinality(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                             const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const;
+
+  /**
+   * Returns whether a given predicate type and its parameter(s) can be pruned.
+   * This method is specialized for strings to handle predicates uniquely applicable to string columns.
+   */
+  bool can_prune(const PredicateCondition predicate_type, const AllTypeVariant& variant_value,
+                 const std::optional<AllTypeVariant>& variant_value2 = std::nullopt) const override;
+
+  /**
+   * Returns the lower bound (minimum value) of the histogram.
+   * This is equal to the smallest value in the segment.
+   */
+  T min() const;
+
+  /**
+   * Returns the upper bound (maximum value) of the histogram.
+   * This is equal to the largest value in the segment.
+   */
+  T max() const;
+
+  /**
+   * Returns the number of bins actually present in the histogram.
+   * This number can differ from the number of bins requested when creating a histogram.
```

Comment:
why?

Reply:
This was only explained in the functions generating the bins thus far.

Changed to:

```c++
/**
   * Returns the number of bins actually present in the histogram.
   * This number can be smaller than the number of bins requested when creating a histogram.
   * See implementations of this method in specific histograms for details.
   */
```

EqualWidth:

```c++
/**
   * Returns the number of bins actually present in the histogram.
   * This number can be smaller than the number of bins requested when creating a histogram.
   * The number of bins is capped at the number of possible values in the range of the segment for integers and strings.
   * That is, if the minimum and maximum value in an integer segment are 1 and 100, there will be at most 100 bins.
   * Otherwise, there would be values which belonged to more than one bin.
   * This is a theoretical problem for floating point numbers as well, but not solved,
   * because the number of representable floating point numbers in a range is not trivial to determine.
   */
```

EqualHeight / EqualElementCount

```c++
/**
   * Returns the number of bins actually present in the histogram.
   * This number can be smaller than the number of bins requested when creating a histogram.
   * The number of bins is capped at the number of distinct values in the segment.
   * Otherwise, there would be empty bins without any benefit.
   */
```

Does that make sense?

Reply:
Yepp